src/extensions/agent_modules/shared/state_store/src/agent_state_store.c

```diff
@@ -513,251 +207,244 @@ ADUC_STATE_STORE_RESULT ADUC_StateStore_SetDeviceId(const char* deviceId)
- * @param instanceName The value to set.
+ * @brief Set the scope Id.
+ * @param scopeId The scope id.
  * @return ADUC_STATE_STORE_RESULT_OK on success, ADUC_STATE_STORE_RESULT_ERROR on failure.
  */
-ADUC_STATE_STORE_RESULT ADUC_StateStore_SetDeviceUpdateServiceInstance(const char* instanceName)
+ADUC_STATE_STORE_RESULT ADUC_StateStore_SetScopeId(const char* scopeId)
 {
+    ADUC_STATE_STORE_RESULT result = ADUC_STATE_STORE_RESULT_ERROR;
+
     sem_wait(&state_semaphore);
-    free(state.deviceUpdateServiceInstance);
-    state.deviceUpdateServiceInstance = SafeStrdup(instanceName);
+
+    free(state.scopeId);
+    state.scopeId = SafeStrdup(scopeId);
+    if (state.scopeId != NULL)
+    {
+        result = ADUC_STATE_STORE_RESULT_OK;
+    }
+
     sem_post(&state_semaphore);
-    return ADUC_STATE_STORE_RESULT_OK;
+    return result;
 }
 
-bool ADUC_StateStore_GetIsDeviceEnrolled(void)
+/**
+ * @brief Whether the scoped or the non-scoped topic is currently subscribed to.
+ *
+ * @param topic The topic in question.
+ * @param isScoped Whether this is the single scoped topic, or the single non-scoped topic.
+ * @return ADUC_STATE_STORE_TOPIC_SUBSCRIBE_STATUS The topic subscribe status.
+ */
+bool ADUC_StateStore_GetTopicSubscribedStatus(const char* topic, bool isScoped)
 {
+    bool result = false;
     sem_wait(&state_semaphore);
-    bool value = state.isDeviceEnrolled;
+
+    if (isScoped)
+    {
+        if (state.scopedTopic != NULL && strncmp(state.scopedTopic, topic, MAX_ADU_MQTT_TOPIC_BYTE_LEN) == 0)
+        {
+            result = true;
+        }
+    }
+    else
+    {
+        if (state.nonscopedTopic != NULL && strncmp(state.nonscopedTopic, topic, MAX_ADU_MQTT_TOPIC_BYTE_LEN) == 0)
+        {
+            result = true;
+        }
+    }
+
     sem_post(&state_semaphore);
-    return value;
+    return result;
 }
 
-ADUC_STATE_STORE_RESULT ADUC_StateStore_SetIsDeviceEnrolled(bool isDeviceEnrolled)
+/**
+ * @brief Sets the subscribe status of the scoped or the non-scoped topic.
+ *
+ * @param topic The topic.
+ * @param isScoped True to set the subscribed status of the single scoped topic. False to set the subscribed status of the single non-scoped topic.
+ * @param newStatus The new subscribe status for the topic.
+ * @return ADUC_STATE_STORE_RESULT_OK on successful setting of subscribe/unsubscribe status.
+ * @return ADUC_STATE_STORE_RESULT_ERROR_EMPTY_TOPIC if the topic is an empty or null string.
+ * @return ADUC_STATE_STORE_RESULT_ERROR_MAX_TOPIC_BYTE_LENGTH_EXCEEDED if the utf-8 topic string's byte length exceeds MAX_ADU_MQTT_TOPIC_BYTE_LEN
+ * @return ADUC_STATE_STORE_RESULT_UNKNOWN_TOPIC when trying to unsubscribe from the scoped or non-scoped topic that is not currently subscribed.
+ * @return ADUC_STATE_STORE_RESULT_ERROR when failed subscribe/unsubscribe an existing topic for some other reason.
+ */
+ADUC_STATE_STORE_RESULT ADUC_StateStore_SetTopicSubscribedStatus(const char* topic, bool isScoped, bool subscribed)
```

Comment:
is there a call to this function that would never have subscribed set to true?

Reply:
Good point. Not currently because in ADUC_Communication_Channel_OnSubscribe() it ignores the reason code--if it's a failure reason code, then it should call this with false so that it's not subscribed because the SUBACK had a failure.

Opened this bug on it to track fixing it:

[Bug 48162686](https://microsoft.visualstudio.com/OS/_workitems/edit/48162686): [Gen2] ADUC_Communication_Channel_OnSubscribe needs to check mqtt reason code of SUBACK to know when topic subscribe failed